File: library/kotlin/src/io/envoyproxy/envoymobile/stats/Tag.kt
package io.envoyproxy.envoymobile

import java.util.regex.Pattern

/**
 * Tag for a stats.
 *
 * Element values must conform to the [Tag.TAG_REGEX].
 */

class Tag(internal val name: String, internal val value: String) {
  init {
    require(TAG_PATTERN.matcher(name).matches()) {
      "Tag names must conform to the regex $TAG_REGEX"
    }
    require(TAG_PATTERN.matcher(value).matches()) {
      "Tag values must conform to the regex $TAG_REGEX"
    }
  }
  
  companion object {
    private const val TAG_REGEX = "^({){0,1}[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}(}){0,1}"
    private val TAG_PATTERN = Pattern.compile(TAG_REGEX)
  }
}

File: library/swift/stats/Tag.swift
import Foundation

/// Tag represents the [key,value] component of a time series name.
/// Element values must conform to the regex /^[A-Za-z_]+$/.
@objcMembers
public final class Tag: NSObject {
  internal let key: String
  internal let value: String

  public init(key: String, value: String) {
    self.key = key
    self.value = value
  }

  public override func isEqual(_ object: Any?) -> Bool {
    return (object as? Tag)?.key == self.key && (object as? Tag)?.value == self.value
  }
}
